drivers/clock_control/clock_control_nrf2_audiopll.c

```diff
@@ -0,0 +1,309 @@
+#include <nrfs_backend_ipc_service.h>
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_DECLARE(clock_control_nrf2, CONFIG_CLOCK_CONTROL_LOG_LEVEL);
+
+#define SHIM_MIN_FREQUENCY 10666667
+#define SHIM_MAX_FREQUENCY 13333293
+
+#define SHIM_DEFAULT_PRESCALER AUDIOPLL_DIV_4
+
+BUILD_ASSERT(
+	DT_NUM_INST_STATUS_OKAY(DT_DRV_COMPAT) == 1,
+	"multiple instances not supported"
+);
+
+BUILD_ASSERT(DT_INST_PROP(0, frequency) >= SHIM_MIN_FREQUENCY);
+BUILD_ASSERT(DT_INST_PROP(0, frequency) <= SHIM_MAX_FREQUENCY);
+
+struct shim_data {
+	struct onoff_manager mgr;
+	onoff_notify_fn mgr_notify;
+	const struct device *dev;
+	struct k_sem evt_sem;
+	nrfs_audiopll_evt_type_t evt;
+};
+
+static int shim_nrfs_request_enable(const struct device *dev)
+{
+	struct shim_data *dev_data = dev->data;
+	nrfs_err_t err;
+
+	LOG_DBG("send enable request");
+
+	dev_data->evt = NRFS_AUDIOPLL_EVT_ENABLED;
+	err = nrfs_audiopll_enable_request(dev_data);
+	if (err != NRFS_SUCCESS) {
+		return -EIO;
+	}
+
+	return 0;
+}
+
+static int shim_nrfs_request_disable(const struct device *dev)
+{
+	struct shim_data *dev_data = dev->data;
+	nrfs_err_t err;
+
+	LOG_DBG("send disable request");
+
+	dev_data->evt = NRFS_AUDIOPLL_EVT_DISABLED;
+	err = nrfs_audiopll_disable_request(dev_data);
+	if (err != NRFS_SUCCESS) {
+		return -EIO;
+	}
+
+	return 0;
+}
+
+static void onoff_start_option(struct onoff_manager *mgr, onoff_notify_fn notify)
+{
+	struct shim_data *dev_data = CONTAINER_OF(mgr, struct shim_data, mgr);
+	const struct device *dev = dev_data->dev;
+	int ret;
+
+	dev_data->mgr_notify = notify;
+
+	ret = shim_nrfs_request_enable(dev);
+	if (ret) {
+		dev_data->mgr_notify = NULL;
+		notify(mgr, -EIO);
+	}
+}
+
+static void onoff_stop_option(struct onoff_manager *mgr, onoff_notify_fn notify)
+{
+	struct shim_data *dev_data = CONTAINER_OF(mgr, struct shim_data, mgr);
+	const struct device *dev = dev_data->dev;
+	int ret;
+
+	dev_data->mgr_notify = notify;
+
+	ret = shim_nrfs_request_disable(dev);
+	if (ret) {
+		dev_data->mgr_notify = NULL;
+		notify(mgr, -EIO);
+	}
+}
+
+static const struct onoff_transitions shim_mgr_transitions = {
+	.start = onoff_start_option,
+	.stop  = onoff_stop_option
+};
+
+/*
+ * Formula:
+ *
+ *   frequency = ((4 + (freq_fraction * 2^-16)) * 32000000) / 12
+ *
+ * Simplified linear approximation:
+ *
+ *   frequency = 10666667 + (((13333293 - 10666667) / 65535) * freq_fraction)
+ *   frequency = 10666667 + ((2666626 / 65535) * freq_fraction)
+ *   frequency = ((10666667 * 65535) + (2666626 * freq_fraction)) / 65535
+ *   frequency = (699040021845 + (2666626 * freq_fraction)) / 65535
+ *
+ * Isolate freq_fraction:
+ *
+ *   frequency = (699040021845 + (2666626 * freq_fraction)) / 65535
+ *   frequency * 65535 = 699040021845 + (2666626 * freq_fractio)
+ *   (frequency * 65535) - 699040021845 = 2666626 * freq_fractio
+ *   freq_fractio = ((frequency * 65535) - 699040021845) / 2666626
+ */
+static uint16_t shim_frequency_to_freq_fraction(uint32_t frequency)
```

Comment:
Should there be a guard/assert here for good measure?
if ``frequency` is too low -> on line 130 freq_fraction will become negative and wrap.

Reply:
The only frequency set is the fixed one in the devicetree, so the `BUILD_ASSERT()`s take care of it :) There was a runtime check in an earlier version of the driver which validated the frequency before passing it to the conversion function since it could vary runtime :)